pkg/generators/apiLinter.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/gengo/types"
+	"k8s.io/kube-openapi/pkg/generators/rules"
+)
+
+// CustomArgs is used by the gengo framework to pass args specific to the API linter
+type CustomArgs struct {
+	ReportFilename string
+}
+
+type apiLinter struct {
+	rules          []APIRule
+	violations     []apiViolation
+	reportFilename string
+}
+
+func newAPILinter(reportFilename string) *apiLinter {
+	return &apiLinter{
+		rules: []APIRule{
+			&rules.NamesMatch{},
+		},
+		reportFilename: reportFilename,
+	}
+}
+
+type apiViolation struct {
+	// Name of rule from APIRule.Name()
+	rule string
+
+	packageName string
+	typeName    string
+
+	// Optional: name of field that violates API rule. Empty fieldName implies that
+	// the entire type violates the rule.
+	field string
+}
+
+// APIRule is the interface for validating API rule on Go types
+type APIRule interface {
+	// Validate evaluates API rule on type t and returns a list of field names in
+	// the type that violate the rule. Empty field name [""] implies the entire
```

Comment:
Is there something better than empty field we can use?  Maybe the name of the type?

Reply:
Will this only lint fields, what about packages or x-package issues?

Reply:
> Is there something better than empty field we can use? Maybe the name of the type?

We may have field's name same as the name of the type. However as you pointed out in the other comments, empty Go name with specified JSON name is considered violation. 

Since the names_match rule is field-level, we could use name of the type or empty string to indicate certain field (people still need to look at source code to figure out if there are two empty Go name violations). I would suggest let's keep it this way (package name, type name, optional field name) for now and change it in future if we need more specific struct to identify API violation (e.g. package name, type name, rule-specific ID)

> Will this only lint fields, what about packages or x-package issues?

names_match evaluates single type. An implementation of APIRule could build its own cache to evaluate package and x-package issues.